### DISCUSSION TITLE
Using `$prefix` in Query.php

### DISCUSSION THREAD

**Discussion by spencerfinnell** [berlindb/core#114] [Q&A]
After noticing [`Query::set_prefix()`](https://github.com/berlindb/core/blob/master/src/Database/Query.php#L387-L391) I figured it would be a good idea to define a `Query::$prefix` property in my class that inherits the `Query` base. However, upon doing so I started seeing unexpected results.

After investigating it looks like it's possible to end up with shared cache groups because [`Base::apply_prefix()` will see the `::$prefix` property](https://github.com/berlindb/core/blob/master/src/Database/Base.php#L146-L150) and then try to apply that to the empty passed string which comes from the default empty `Query::$cache_group`. Basically you get `my-prefix-` for every group name.

Explicitly defining the `Query::$cache_group` property resolved my issue but I'm wondering if there is a good place to document that both are needed if used, or if the parent class should throw an exception if there is a set prefix with no cache group?

**Comment by JJJ** [comment]
This variable detail definitely should be made better soon. It's pretty confusing.

The `::$prefix` property is _intended_ to be set once in the `Base` class for everything that extends it, and I did not plan on it being set differently by subclasses.

For [Sugar Calendar](https://github.com/sugarcalendar/core/blob/7155c47aa1925c35ec33729ee55d31ba6629d21e/sugar-calendar/includes/classes/database/engine/class-base.php#L44-L50) the `Base::prefix` gets set to `sc` and for [EDD](https://github.com/easydigitaldownloads/easy-digital-downloads/blob/8128b5383cd3610778da33e27b8aa35f8b797701/includes/database/engine/class-base.php#L44-L50) it's `edd`.

Obviously though, this requires hard-coding a change into Berlin because you can't extend 2 classes at once. See #103.

Would love ideas on how to make this super easy.

**Reply by alexstandiford** [reply]
I wonder if simply using __set would be sufficient? Make it so that it can only be set once, perhaps?

**Comment by JJJ** [comment]
Fixed for 3.1.0 via #195.

Thanks everyone 🙏
